enhancements/storage/csi-migration.md

```diff
@@ -0,0 +1,220 @@
+---
+title: csi-migration
+authors:
+  - "@fbertina"
+reviewers:
+  - "@openshift/storage”
+approvers:
+  - "@openshift/openshift-architects"
+creation-date: 2020-07-29
+last-updated: 2021-03-05
+status: provisional
+see-also: https://github.com/openshift/enhancements/pull/463
+replaces:
+superseded-by:
+---
+
+# Migration of in-tree volume plugins to CSI drivers
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+We want to allow cluster administrators to seamlessly migrate volumes created using the in-tree storage plugin to their counterpart CSI drivers. It is important to achieve this goal before CSI Migration feature becomes GA in upstream. This also a requirement for supporting [out-of-tree cloud providers](https://github.com/openshift/enhancements/pull/463)
+
+## Motivation
+
+[CSI migration](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/625-csi-migration) is an upstream effort to migrate in-tree volume plugins to their counterpart CSI drivers. The feature is beta since Kubernetes 1.17, however, as of Kubernetes 1.20 it is still *disabled* by default.
+
+That is going to change in Kubernetes 1.21 (OCP 4.8), where the feature will remain beta, but *enabled* by default. In Kubernetes 1.22 (OCP 4.9) the feature may become GA.
+
+In OCP we can optionally disable the CSI migration feature while it is still beta, however, that will no longer be an option once CSI migration becomes GA. In order to avoid surprises, once the migration is enabled by default in OCP, we want to allow cluster administrators to optionally enable the feature earlier, preferably in OCP 4.8.
+
+### Goals
+
+Our goals are different throughout our support lifecycle.
+
+For Tech Preview, we want to introduce a mechanism to allow switching CSI migration feature flags on and off across OCP components. It is important that this mechanism allows for a seamless migration path, without breaking existing volumes.
+
+For GA, existing in-tree volumes will be migrated to use the CSI driver. Once the migration has been enabled, users should not have to do any additional work. In this phase we will not support disabling CSI migration, and we should ensure that downgrades are not broken.
+
+### Non-Goals
+
+* Control the ordering in which OCP components will be upgraded or downgraded.
+* Install or remove the CSI driver as the migration is enabled or disabled.
+
+## Proposal
+
+We propose to add a carry-patch to the Attach Detach Controller in OCP that enables the migration of some storage plugins. Initially we would start with Cinder and GCP, so that we are aligned with the goals of [CCMO](https://github.com/openshift/enhancements/pull/463).
```

Comment:
@bertinatto can you please add a sentence that we propose to use FeatureSet `TechPreviewNoUpgrade` to trigger the migration during tech preview phase and it will enable cloud migration too (with a link to their enhancement)?

(somewhere below in tech details we could mention that it's possible to enable CSI migration without cloud migration using 
```
apiVersion: config.openshift.io/v1
kind: FeatureGate
metadata:
  name: cluster
spec:
  featureSet: CustomNoUpgrade
  customNoUpgrade:
    enabled:
    - CSIMigrationAWS
```
)

Reply:
Also added examples for `CustomNoUpgrade` and `TechPreviewNoUpgrade` in  the `Graduation` section.